### PR TITLE
feat: enhance backup

### PR DESCRIPTION
**Pull request by iskdrews**
## Explanation

Refactoring Backup edge cases. 

## More Information

- Related to #257 

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

> PR template source from [github.com/MetaMask](https://github.com/MetaMask)

### CODE REVIEW

**Review by AtHeartEngineer — APPROVED** [review]

File: src/background/services/approval/index.ts [hunk 1]
```diff
@@ -97,8 +97,8 @@ export default class ApprovalService implements IBackupable {
 
   downloadEncryptedStorage = (): Promise<string | null> => this.approvals.get<string>();
 
-  uploadEncryptedStorage = async (encryptedApprovals: string, password: string): Promise<void> => {
-    await this.lockService.checkPassword(password);
-    await this.approvals.set(encryptedApprovals);
+  uploadEncryptedStorage = async (encryptedBackup: string, password: string): Promise<void> => {
+    const { authenticBackupCiphertext } = await this.lockService.isAuthenticated(password, encryptedBackup);
+    if (authenticBackupCiphertext) await this.approvals.set(authenticBackupCiphertext);
```
Comment (0xmad) [inline src/background/services/approval/index.ts:102]: Maybe we need throw error if backup is corrupted?

File: src/background/services/crypto/index.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export function isCryptoHmacAuthentic(ciphertext: string, password: string): boolean {
+  const { transitHmac, transitCipherContent } = cryptoSubHmacCiphertext(ciphertext);
+
+  const decryptedHmac = generateHmac(transitCipherContent, password);
+
+  if (transitHmac !== decryptedHmac) return false;
+
+  return true;
```
Comment (0xmad) [inline src/background/services/crypto/index.ts:29]:
```suggestion
  return transitHmac === decryptedHmac;
```

File: src/background/services/crypto/index.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export function cryptoDecrypt(ciphertext: string, password: string): string {
+  const bytes = AES.decrypt(ciphertext, password);
+  return bytes.toString(enc.Utf8);
+}
+
+export function generateEncryptedHmac(ciphertext: string, password: string): string {
+  const hmac = generateHmac(ciphertext, password);
+  return hmac + ciphertext;
```
Comment (0xmad) [inline src/background/services/crypto/index.ts:15]:
```suggestion
  return `${hmac}${ciphertext}`;
```

File: src/background/services/zkIdentity/index.ts [hunk 4]
```diff
@@ -357,8 +356,8 @@ export default class ZkIdentityService implements IBackupable {
 
   downloadEncryptedStorage = async (): Promise<string | null> => this.identitiesStore.get<string>();
 
-  uploadEncryptedStorage = async (encrypted: string, password: string): Promise<void> => {
-    await this.lockService.checkPassword(password);
-    await this.identitiesStore.set<string>(encrypted);
+  uploadEncryptedStorage = async (encryptedBackup: string, password: string): Promise<void> => {
+    const { authenticBackupCiphertext } = await this.lockService.isAuthenticated(password, encryptedBackup);
+    if (authenticBackupCiphertext) await this.identitiesStore.set<string>(authenticBackupCiphertext);
```
Comment (0xmad) [inline src/background/services/zkIdentity/index.ts:361]: Same here with error case

File: src/background/services/lock/index.ts [hunk 5]
```diff
@@ -99,27 +103,58 @@ export default class LockService implements IBackupable {
...
+    backupCiphertext?: string,
+  ): Promise<{ isAuthentic: boolean; isBackup?: boolean; isNew?: boolean; authenticBackupCiphertext?: string }> => {
     const cipherText = await this.passwordStorage.get<string>();
 
-    if (!cipherText) {
-      throw new Error("Something badly gone wrong (reinstallation probably required)");
+    if (!password) throw new Error("Password is not provided");
+
+    // Normal check
+    if (cipherText) {
```
Comment (0xmad) [inline src/background/services/lock/index.ts:121]: It's better to get rid of nested conditions

File: src/background/services/lock/index.ts [hunk 6]
```diff
@@ -99,27 +103,58 @@ export default class LockService implements IBackupable {
...
+      const decryptedPasswordChecker = cryptoDecrypt(cipherText, password);
+
+      if (decryptedPasswordChecker !== this.passwordChecker) throw new Error("Incorrect password");
+      else {
+        if (backupCiphertext)
+          return {
+            isAuthentic: true,
+            isNew: false,
+            authenticBackupCiphertext: this.isBackupCiphertextAutentic(backupCiphertext),
```
Comment (0xmad) [inline src/background/services/lock/index.ts:130]: You can move it to a variable and get rid of code duplication

File: src/background/services/lock/index.ts [hunk 7]
```diff
@@ -99,27 +103,58 @@ export default class LockService implements IBackupable {
...
 
-    const bytes = CryptoJS.AES.decrypt(cipherText, password);
-    const retrievedPasswordChecker = bytes.toString(CryptoJS.enc.Utf8);
-
-    if (retrievedPasswordChecker !== this.passwordChecker) {
-      throw new Error("Incorrect password");
+  private isBackupCiphertextAutentic = (backupCiphertext: string): string => {
+    const { isHmacAuthentic, authenticCiphertext: authenticBackupCiphertext } = this.isHmacAuthentic(backupCiphertext);
+    if (!isHmacAuthentic || !authenticBackupCiphertext) throw new Error("This backup file is not authentic.");
+    else {
```
Comment (0xmad) [inline src/background/services/lock/index.ts:156]: No need in else case here you can just return the value

File: src/background/services/lock/index.ts [hunk 8]
```diff
@@ -99,27 +103,58 @@ export default class LockService implements IBackupable {
...
+      throw new Error("Something badly gone wrong (reinstallation probably required)");
     }
+  };
 
-    const bytes = CryptoJS.AES.decrypt(cipherText, password);
-    const retrievedPasswordChecker = bytes.toString(CryptoJS.enc.Utf8);
-
-    if (retrievedPasswordChecker !== this.passwordChecker) {
-      throw new Error("Incorrect password");
+  private isBackupCiphertextAutentic = (backupCiphertext: string): string => {
```
Comment (0xmad) [inline src/background/services/lock/index.ts:153]: Methods which starts with `is` is supposed to return boolean value. I'd rename it to `getBackupCiphertextAutentic`

File: src/background/services/lock/index.ts [hunk 9]
```diff
@@ -132,20 +167,25 @@ export default class LockService implements IBackupable {
...
-    if (!this.password) {
-      throw new Error("Password is not provided");
-    }
+    if (!this.password) throw new Error("Password is not provided");
+    return cryptoDecrypt(ciphertext, this.password);
+  };
 
-    const bytes = CryptoJS.AES.decrypt(ciphertext, this.password);
-    return bytes.toString(CryptoJS.enc.Utf8);
+  private isHmacAuthentic = (ciphertext: string): { isHmacAuthentic: boolean; authenticCiphertext?: string } => {
```
Comment (0xmad) [inline src/background/services/lock/index.ts:179]: Same here with method name, I'd rename it to `getHmacAuthenticData`

**Review by 0xmad — CHANGES_REQUESTED** [review]
@iskdrews thanks, there are just some refactoring suggestions and design improvements.

File: src/background/services/lock/index.ts [hunk 10]
```diff
@@ -97,30 +102,64 @@ export default class LockService implements IBackupable {
...
+
+    const cipherText = await this.passwordStorage.get<string>();
+    if (cipherText) {
+      const decryptedPasswordChecker = cryptoDecrypt(cipherText, password);
+
+      return decryptedPasswordChecker === this.passwordChecker;
     }
+
+    return false;
   };
```
Comment (0xmad) [inline src/background/services/lock/index.ts:163]:
```suggestion
  private isLockerPasswordAuthentic = async (password: string): Promise<boolean> => {
    if (!password) throw new Error("Password is not provided");

    const cipherText = await this.passwordStorage.get<string>();
    
    if (!cipherText) {
      return false;
    }
    
    const decryptedPasswordChecker = cryptoDecrypt(cipherText, password);
    return decryptedPasswordChecker === this.passwordChecker;
  };
```

File: src/background/services/lock/index.ts [hunk 11]
```diff
@@ -97,30 +102,64 @@ export default class LockService implements IBackupable {
...
+    return true;
+  };
 
-    if (retrievedPasswordChecker !== this.passwordChecker) {
-      throw new Error("Incorrect password");
+  private isNewOnboarding = async () => {
+    const cipherText = await this.passwordStorage.get<string>();
+    if (!cipherText) return true;
+    return false;
+  };
```
Comment (0xmad) [inline src/background/services/lock/index.ts:150]:
```suggestion
  private isNewOnboarding = async () => {
    const cipherText = await this.passwordStorage.get<string>();
    
    return !cipherText;
  };
```

File: src/background/services/lock/index.ts [hunk 12]
```diff
@@ -28,27 +30,30 @@ export default class LockService implements IBackupable {
...
 
+  private backupService: BackupService;
+
   private constructor() {
     this.isUnlocked = false;
     this.passwordChecker = "Password is correct";
     this.passwordStorage = new SimpleStorage(PASSWORD_DB_KEY);
     this.password = undefined;
     this.unlockCB = undefined;
+    this.backupService = BackupService.getInstance();
```
Comment (0xmad) [inline src/background/services/lock/index.ts:41]: I don't think it's correct to use backup service inside backupable. You can move `getAuthenticBackupCiphertext` method to crypto service and import here so with don't have cycle dependency between services.

File: src/background/services/approval/index.ts [hunk 13]
```diff
@@ -16,12 +18,15 @@ export default class ApprovalService implements IBackupable {
...
+  private lockService: LockerService;
+
+  private backupService: BackupService;
 
   private constructor() {
     this.allowedHosts = new Map();
     this.approvals = new SimpleStorage(APPPROVALS_DB_KEY);
-    this.lockService = LockService.getInstance();
+    this.lockService = LockerService.getInstance();
+    this.backupService = BackupService.getInstance();
```
Comment (0xmad) [inline src/background/services/approval/index.ts:29]: The same here for backup service and backupable

File: src/background/services/zkIdentity/index.ts [hunk 14]
```diff
@@ -43,10 +46,11 @@ export default class ZkIdentityService implements IBackupable {
     this.activeIdentity = undefined;
     this.identitiesStore = new SimpleStorage(IDENTITY_KEY);
     this.activeIdentityStore = new SimpleStorage(ACTIVE_IDENTITY_KEY);
-    this.lockService = LockService.getInstance();
+    this.lockService = LockerService.getInstance();
     this.notificationService = NotificationService.getInstance();
     this.historyService = HistoryService.getInstance();
     this.browserController = BrowserUtils.getInstance();
+    this.backupService = BackupService.getInstance();
```
Comment (0xmad) [inline src/background/services/zkIdentity/index.ts:53]: And the last with backup service and backupable

**Review by 0xmad — APPROVED** [review]
@iskdrews thanks!
